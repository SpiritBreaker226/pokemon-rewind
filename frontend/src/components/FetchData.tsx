import React, { FunctionComponent, useContext, useEffect } from 'react'

import axios from 'axios'

import { Typography } from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'

import { AppContext } from '../contexts/AppContext'

import { Types } from '../types/Actions'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    message: {
      margin: theme.spacing(10),
      textAlign: 'center',
    },
  })
)

const FetchData: FunctionComponent = () => {
  const classes = useStyles()
  const { state, dispatch } = useContext(AppContext)

  useEffect(() => {
    const fetchData = async () => {
      try {
        dispatch({
          type: Types.UpdateLoading,
          payload: { isLoading: true },
        })

        const res = await axios.get(state.urlToEndpoint)

        dispatch({
          type: Types.UpdateCards,
          payload: {
            cards: res.data.cards.data,
            pagination: res.data.pagination,
            isLoading: false,
          },
        })
      } catch (error) {
        dispatch({
          type: Types.UpdateErrorMessageFromServer,
          payload: {
            isLoading: false,
            errorMessage: error.message,
          },
        })
      }
    }

    fetchData()
  }, [dispatch, state.urlToEndpoint])

  if (state.isLoading)
    return (
      <div className={classes.message}>
        <Typography variant="h6">Loading...</Typography>
      </div>
    )
  if (state.errorMessage) {
    return (
      <div className={classes.message}>
        <Typography variant="h6" color="error">
          {state.errorMessage}
        </Typography>
      </div>
    )
  }

  return null
}

export default FetchData
