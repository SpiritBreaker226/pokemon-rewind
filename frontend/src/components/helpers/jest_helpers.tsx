import React, { FunctionComponent } from 'react'
import UserEvent from '@testing-library/user-event'
import { within, waitForElementToBeRemoved } from '@testing-library/react'

import { Card } from '../../types/Cards'
import { Direction, Method } from '../../types/CardQuery'
import { Pagination } from '../../types/Pagination'
import { Action } from '../../types/Actions'
import { Search } from '../../types/Search'

import { AppContext } from '../../contexts/AppContext'

export const cards: Card[] = [
  {
    attributes: {
      name: 'Lightning Energy',
      supertype: 'Energy',
      subtype: 'Basic',
      number: '128',
      artist: 'Keiji Kinebuchi',
      rarity: 'Rare',
      series: 'Base',
      set_code: 'base4',
    },
  },
  {
    attributes: {
      name: 'Jigglypuff',
      national_pokedex_number: 39,
      supertype: 'Pokémon',
      subtype: 'Basic',
      hp: 60,
      converted_retreat_cost: 1,
      number: '77',
      artist: 'Kagemaru Himeno',
      rarity: 'Common',
      series: 'Base',
      set_code: 'base4',
    },
  },
  {
    attributes: {
      name: 'Squirtle',
      national_pokedex_number: 7,
      supertype: 'Pokémon',
      subtype: 'Basic',
      hp: 40,
      converted_retreat_cost: 1,
      number: '93',
      artist: 'Mitsuhiro Arita',
      rarity: 'Common',
      series: 'Base',
      set_code: 'base4',
    },
  },
]

export interface TestingState {
  cards?: Card[]
  pagination?: Pagination
  urlToEndpoint?: string
  search?: Search
  method?: Method
  errorMessage?: string
  isLoading?: boolean
  sorting?: {
    fieldName: string | undefined
    dir: Direction
  }
}

export interface MakeWrapperProps {
  state: TestingState
  dispatch?: (action: Action) => void
}

export const MakeWrapper: FunctionComponent<MakeWrapperProps> = ({
  state,
  dispatch,
  children,
}) => (
  <AppContext.Provider
    value={{
      state: {
        cards: [],
        pagination: {
          rows_per_page: 10,
          page_total: 0,
          current_page: '1',
        },
        urlToEndpoint: '',
        search: {
          toggle: false,
          value: {
            name: '',
            hp: '',
            rarity: '',
          },
          field: 'name',
        },
        method: 'GET',
        errorMessage: '',
        sorting: {
          fieldName: undefined,
          dir: undefined,
        },
        isLoading: false,
        ...state,
      },
      dispatch: dispatch || (() => null),
    }}
  >
    {children}
  </AppContext.Provider>
)

export const selectMaterialUiSelectOption = async (
  selectElement: HTMLElement,
  optionText: string
) =>
  new Promise((resolve) => {
    if (!selectElement || !selectElement.parentNode) {
      throw new Error(
        'Unable to find select element for selectMaterialUiSelectOption'
      )
      return
    }

    // The the button that opens the dropdown, which is a sibling of the input
    const selectButton = selectElement.parentNode.querySelector('[role=button]')

    if (!selectButton) {
      throw new Error(
        'Unable to find select button for selectMaterialUiSelectOption'
      )
      return
    }

    // Open the select dropdown
    UserEvent.click(selectButton)

    // Get the dropdown element. We don't use getByRole() because it includes <select>s too.
    const listbox = document.body.querySelector<HTMLElement>('ul[role=listbox]')

    if (!listbox) {
      throw new Error('Unable to find listbox for selectMaterialUiSelectOption')
      return
    }

    // Click the list item
    const listItem = within(listbox).getByText(optionText)
    UserEvent.click(listItem)

    // Wait for the listbox to be removed, so it isn't visible in subsequent calls
    waitForElementToBeRemoved(() =>
      document.body.querySelector('ul[role=listbox]')
    ).then(resolve)
  })
