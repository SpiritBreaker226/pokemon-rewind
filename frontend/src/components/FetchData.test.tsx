import React from 'react'

import { waitFor, screen, render } from '@testing-library/react'

import axios from 'axios'

import { MakeWrapper } from './helpers/jest_helpers'

import FetchData from './FetchData'

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

describe('FetchData', () => {
  describe('fetch card data', () => {
    it('renders loading screen', async () => {
      render(
        <MakeWrapper state={{ isLoading: true }}>
          <FetchData />
        </MakeWrapper>
      )

      expect(screen.getByText('Loading', { exact: false })).toBeInTheDocument()
      expect(screen.getByText('Loading', { exact: false })).toMatchSnapshot()

      await waitFor(() => {})
    })

    describe('on error from the server', () => {
      it('should render an error message', async () => {
        mockedAxios.get.mockRejectedValue(new Error('fake error message'))

        render(
          <MakeWrapper
            state={{
              errorMessage: 'fake error message',
            }}
          >
            <FetchData />
          </MakeWrapper>
        )

        await waitFor(() => screen.getByText('error', { exact: false }))

        const error = screen.getByText('fake error message', { exact: false })

        expect(error).toBeInTheDocument()
        expect(error).toMatchSnapshot()
      })
    })
  })
})
