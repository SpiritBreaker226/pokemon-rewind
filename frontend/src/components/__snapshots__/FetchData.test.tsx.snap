// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchData fetch card data on error from the server should render an error message 1`] = `
<h6
  class="MuiTypography-root MuiTypography-h6 MuiTypography-colorError"
>
  fake error message
</h6>
`;

exports[`FetchData fetch card data renders loading screen 1`] = `
<h6
  class="MuiTypography-root MuiTypography-h6"
>
  Loading...
</h6>
`;
